### ISSUE TITLE
libssh2_sftp_read() with maxlen of 8192 causes memory corruption

### ISSUE THREAD

**Issue by Corillian** [libssh2/libssh2#50]
When calling `libssh2_sftp_read()` with a buffer size of 4096 my application works as expected. If I do nothing other than change the buffer size to 8192 the application begins crashing due to memory corruption within libssh2. I am using a blocking socket with libssh2 set to blocking mode. Problems start occurring when libssh2 arbitrarily skips ahead within the file. Sometimes it does it with the first read and sometimes it doesn't do it until after multiple successful reads:
```
NOTE: Requested size is always 8192

GetBuffer() [New] About to read from requested file position 0 and actual file position 0
GetBuffer() [New] Read completed: bytes read = 2768, totalByteRead = 2768, new actual file position = 10960
GetBuffer() [New] About to read from requested file position 0 and actual file position 10960
GetBuffer() [New] Read completed: bytes read = 5424, totalByteRead = 8192, new actual file position = 16384
```
If you look at the log output above you can see that the libssh2 file pointer position started at 0. 8192 bytes were requested, 2768 were returned (it's always this number once problems begin), and the new file  pointer after read has somehow jumped to **10960**.

I compiled libssh2 and all of its dependencies as 64-bit DLL's using Visual Studio 2013. It's linking to OpenSSL 1.0.2d and zlib 1.2.8. All of my testing was done with Windows 7 x64. My sequence of actions is:
1.  I set function callbacks with OpenSSL for creating and destroying locks
2. Initialize winsock and libssh2
3. Create a (blocking) socket
4. Do the ssh handshake using a private key in memory
5. Create the sftp session and open a file. When the sftp session is created I set function pointers to a custom set of memory allocation functions that just call `malloc()`, `realloc()`, and `free()`
6. Begin reading from the file. Every call to `libssh2_sftp_read()` is preceded by `libssh2_sftp_seek64()` to explicitly set the file pointer

I can't provide a callstack because the memory corruption appears to be nuking it. All I get is an access violation at `msvcr120.dll!000007fedea9c3f9() Unknown`. If there's any additional information I can provide please let me know.

**Comment by sbredahl** [comment]
I can confirm that some weird things happen if the buffer maxlen is much larger than 4k (7k works for me also, but 8k is too large). I end up in an infinite loop where libssh2_sftp_read() keeps returning 0 (in blocking mode). 

I don't get any access violation however (also Win7 x64).

**Comment by sbredahl** [comment]
I have investigated this issue a little further (using example-sftp) on both of my libssh2 builds (Ubuntu 14.10 and Windows 7) at work. 

Corrupted downloads happen (in both builds) if buffer_maxlen * 4 > MAX_SFTP_READ_SIZE. So with the current MAX_SFTP_READ_SIZE=30000, problems occur if buffer_maxlen>7500.

**Comment by jakob** [comment]
@Corillian Could you check if my branch https://github.com/jakob/libssh2/tree/sftp-read-fix (see pull request #75) fixes your issues?

**Comment by bagder** [comment]
That fix is now merged into master and we believe this bug is fixed. Please provide details if not.
